docs/source/debugging.rst

```diff
@@ -0,0 +1,91 @@
+
+When a model with an BoundedConstraintSet is solved, it checks whether any variables slid off to the bounds, notes this in the solution dictionary and prints a warning (if verbosity is greater than 0).
+
+For example, Mosek returns ``DUAL_INFEAS_CER`` when attempting to solve the following model:
+
+.. literalinclude:: examples/unbounded.py
+
+Upon viewing the printed output,
+
+.. literalinclude:: examples/unbounded_output.txt
+
+it becomes clear that the problem is, unsurprisingly, that ``x`` is unbounded below in the original model.
+
+Details
+-------
+
+A dual feasible solution typically means that more than one unique solution meets the objective.   An example of a dual-feasible model is shown below. This model is dual-infeasible because there are multiple values of ``x`` and ``y`` that satisfy the constraint set and yield the globally optimum cost of 0.5.
+
+.. literalinclude:: examples/subinplace.py
+
+``cvxopt`` and ``Mosek`` both solve the above model and output a cost of 0.5, however, the values of ``x`` and ``y`` will be different, illustrating how the model is dual feasible.
+
+The following is an example of a dual-infeasible problem. While the difference is slight, this cannot be solved by either ``mosek`` or ``cvxopt``.  ``cvxopt`` will again give a ``Rank`` error.  ``Mosek`` can identify deal-infeasible models and the error message will label it as such. Typically, this type of error means that one or more variables are not sufficiently bounded.
+
+.. literalinclude:: examples/dual_infeasible_ex2.py
+
+Another common cause of dual-infeasability occurs when a constrain applies pressure on a variable in an unexpected direction and pushs its value to either zero or infinity. When this occurs, Mosek usually returns a final status of dual-infeasible while cvxopt will return a final solver status of unknown. A simple example is given below. ``x`` has no upper bound, and the objective is to minimize ``1/x``, so the solver pushes ``x`` towards infinitiy and returns dual infeasible.
+
+.. literalinclude:: examples/dual_infeasible_ex.py
+
+Debugging large, dual infeasible, models can be difficult. The recommended procedure is to use a ``BoundedConstraintSet``, found in ``gpkit.tools``. ``BoundedConstraintSet`` adds additional constraints to the model that bounds each variable to be greater than or equal to ``eps``, and less than or equal to ``1/eps``. The default value for ``eps`` is 1e-30. This prevents variables from being truly unbounded and allows most dual infeasible models to solve. By inspecting the solution, or by also making use of the a ``Tight Constraint Set``, it is easy to determine which variables are unbounded and modify constraints as necessary. Below, a BoundedConstraintSet is used to make the previous model solvable.
+
+.. literalinclude:: examples/BoundedConstraintSet_ex.py
+
+With the formulation above, ``x`` has a lower bound at 1e-30, so the solver returns a solution with cost 1e-30.
```

Comment:
could be, maybe combine and reorder the sections so that dual infeasible is explained before discussed in debugging?